src/context/EnhancedDataContext.tsx

```diff
@@ -0,0 +1,371 @@
+
+import {
+    calculateSingleTilePosition,
+} from "../mapgen/tilePositioning";
+import { optimizeFactionColors, RGBColor } from "../utils/colorOptimization";
+import { getColorValues } from "../lookup/colors";
+import { createContext, ReactNode } from "react";
+import { usePlayerData } from "@/hooks/usePlayerData";
+import { useParams } from "react-router-dom";
+import { colors } from "@/data/colors";
+import { MapTile, Planet } from "@/types/global";
+import { PlayerDataResponse } from "@/data/types";
+
+
+
+
+export type FactionColorMap = {
+    [key: string]: FactionColorData
+}
+
+export type FactionColorData = {
+    faction: string,
+    color: string,
+    optimizedColor: RGBColor
+}
+
+type EnhancedDataProviderProps = {
+    children: ReactNode;
+};
+
+type EnhancedDataContextValue = {
+    planetAttachments: Record<string, string[]>;
+    mapTiles: MapTile[];
+    tilePositions: any;
+    allExhaustedPlanets: string[];
+    systemIdToPosition: Record<string, string>;
+    factionColorMap: FactionColorMap;
+    tilesWithPds: Set<string>;
+    dominantPdsFaction: Record<
+        string,
+        {
+            faction: string;
+            color: string;
+            count: number;
+            expected: number;
+        }
+    >;
+}
+
+export const EnhancedDataContext = createContext<EnhancedDataContextValue | undefined>(
+    undefined
+);
+
+export function EnhancedDataContextProvider({ children }: EnhancedDataProviderProps) {
+    const params = useParams<{ mapid: string }>();
+    const gameId = params.mapid!;
+    const systemIdToPosition: Record<string, string> = {};
+    const planetAttachments: Record<string, string[]> = {};
+    const factionColorMap: FactionColorMap = {};
+    let validData: PlayerDataResponse;
+
+    const { data, isLoading, isError, refetch } = usePlayerData(gameId);
+
+    if (data) {
+        validData = data!;
+    } else return (<EnhancedDataContext.Provider value={undefined}>
+        {children}
+    </EnhancedDataContext.Provider>);
```

Comment:
two things here

a) keep the data enhancing into a separate method
we should have a method that operates on the data irrespective of react stuff (at some point will want to unit test and will be easier this way)

b) use typescript inference 
```suggestion
    if (!data) {
return (<EnhancedDataContext.Provider value={undefined}>
        {children}
    </EnhancedDataContext.Provider>);
    }
```

no need for `validData`, from here on out typescript will know data is non-nullable